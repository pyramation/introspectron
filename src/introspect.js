import { resolve as resolvePath } from 'path';
import PgConverter from './PgConverter';
import { readFile } from 'fs';

/**
 * The introspection query SQL string. We read this from it’s SQL file
 * synchronously at runtime. It’s just like requiring a file, except that file
 * is SQL.
 */
const introspectionQuery = new Promise((resolve, reject) => {
  readFile(resolvePath(__dirname, './introspect.sql'), (error, data) => {
    if (error) reject(error);
    else resolve(data.toString());
  });
});

/**
 * Takes a Postgres client and introspects it, returning an instance of
 * `PgObjects` which can then be consumed. Note that some translation is done
 * from the raw Postgres catalog to the friendlier `PgObjects` interface.
 */
export default async function introspect(client, schemas) {
  // Run our single introspection query in the database.
  const result = await client.query({
    name: 'introspectionQuery',
    text: await introspectionQuery,
    values: [schemas]
  });

  const objects = result.map(({ object }) => object);

  // Extract out the objects from the query.
  return new PgConverter(objects);
}
