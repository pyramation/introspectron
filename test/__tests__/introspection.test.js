import { getConnection, closeConnection } from '../utils/SQLTest';

import introspect from '../../src/introspect';

let db;
let catalog;

describe('introspect', () => {
  beforeAll(async () => {
    db = await getConnection();
    catalog = await introspect(db, ['test_schema']);
  });
  afterAll(async () => {
    await closeConnection(db);
  });
  describe('converts to JSON schema', () => {
    it('tables', async () => {
      var output = catalog.toTables();
      expect(output).toMatchSnapshot();
    });
    it('procedures', async () => {
      var output = catalog.toProcs();
      expect(output).toMatchSnapshot();
    });
  });
});
