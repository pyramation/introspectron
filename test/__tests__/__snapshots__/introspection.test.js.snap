// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`introspect converts to JSON schema procedures 1`] = `
Object {
  "test_schema/myfunc2": Object {
    "description": "description for function here",
    "properties": Object {
      "_name": Object {
        "type": "text",
      },
    },
    "required": Array [
      "_name",
    ],
    "type": "object",
  },
}
`;

exports[`introspect converts to JSON schema tables 1`] = `
Object {
  "test_schema/example": Object {
    "description": "example table description",
    "properties": Object {
      "a": Object {
        "description": "field a description",
        "type": "int4",
      },
      "b": Object {
        "description": "field b description",
        "type": "int4",
      },
      "c": Object {
        "description": "field c description",
        "type": "int4",
      },
      "name": Object {
        "description": null,
        "type": "text",
      },
      "price": Object {
        "description": null,
        "type": "numeric",
      },
      "product_no": Object {
        "description": null,
        "type": "int4",
      },
      "tags": Object {
        "description": null,
        "items": Object {
          "type": "text",
        },
        "type": "array",
      },
    },
    "required": Array [
      "product_no",
      "a",
      "c",
    ],
    "type": "object",
  },
  "test_schema/products": Object {
    "description": null,
    "properties": Object {
      "a": Object {
        "description": "field a description",
        "type": "int4",
      },
      "b": Object {
        "description": "field b description",
        "type": "int4",
      },
      "c": Object {
        "description": "field c description",
        "type": "int4",
      },
      "name": Object {
        "description": null,
        "type": "text",
      },
      "price": Object {
        "description": null,
        "type": "numeric",
      },
      "product_no": Object {
        "description": null,
        "type": "int4",
      },
      "tags": Object {
        "description": null,
        "items": Object {
          "type": "text",
        },
        "type": "array",
      },
    },
    "required": Array [
      "product_no",
      "a",
      "c",
    ],
    "type": "object",
  },
}
`;
